File: frontend/gui/grpc_client.py
import grpc
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), 'proto'))
import proto.netmap_pb2
import proto.netmap_pb2_grpc

class NetworkClient:
    def __init__(self, channel):
        self.client = proto.netmap_pb2_grpc.NetMapStub(channel)

    def send_network_data(self, network_data):
        request = proto.netmap_pb2.NetworkDataRequest(network_data=network_data)
        response = self.client.SendNetworkData(request)
        print("Response from server:", response.message)
        return response.message

def run():
    # Set up the connection to the Go server
    channel = grpc.insecure_channel('localhost:50051')
    client = NetworkClient(channel)

    # Example network data to send to the server
    network_data = "Requesting current network state"
    client.send_network_data(network_data)

if __name__ == "__main__":
    run()

File: frontend/go_server/grpc_client.go
package main

import (
	"context"
	"fmt"
	"log"
	"net"

	"google.golang.org/grpc"
	"your_project_path/proto" // replace with the correct path to your generated proto files
)

const (
	serverAddress = "localhost:50051" // Python gRPC server address
)

type server struct {
	proto.UnimplementedNetworkServiceServer
}

// Example: A method to send data to the Python UI and get a response
func (s *server) SendNetworkData(ctx context.Context, req *proto.NetworkDataRequest) (*proto.NetworkDataResponse, error) {
	fmt.Println("Received data from client:", req)
	// Process the data if needed, then send the response back
	return &proto.NetworkDataResponse{Message: "Data received and processed!"}, nil
}

func main() {
	// Set up connection to Python gRPC server
	conn, err := grpc.Dial(serverAddress, grpc.WithInsecure(), grpc.WithBlock())
	if err != nil {
		log.Fatalf("Failed to connect to server: %v", err)
	}
	defer conn.Close()

	client := proto.NewNetworkServiceClient(conn)

	// Sending data to Python UI
	req := &proto.NetworkDataRequest{NetworkData: "Current network state"}
	resp, err := client.SendNetworkData(context.Background(), req)
	if err != nil {
		log.Fatalf("Failed to send request: %v", err)
	}

	fmt.Println("Response from Python UI:", resp.GetMessage())
}